src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -127,56 +177,122 @@ def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
+                closures.append(current_closure)
+                continue
+
+            # TODO(tehrengruber): extracting from a scan is untested.
+            is_scan: bool = isinstance(
+                current_closure.stencil, ir.FunCall
+            ) and current_closure.stencil.fun == im.ref("scan")
+            current_closure_stencil = (
+                current_closure.stencil if not is_scan else current_closure.stencil.args[0]  # type: ignore[attr-defined]  # ensured by is_scan
+            )
 
-        stencil_stack = [(closure.output, popped_stencil)]
-        domain = closure.domain
-        while stencil_stack:
-            output, call = stencil_stack.pop()
-            closure = ir.StencilClosure(
-                domain=domain,
-                stencil=call.fun,
-                output=output,
-                inputs=[handle_arg(arg) for arg in call.args],
+            stencil_body, extracted_lifts, _ = extract_subexpression(
+                current_closure_stencil.expr,
+                temporary_extraction_predicate,
+                uid_gen_tmps,
+                once_only=True,
+                deepest_expr_first=True,
             )
-            closures.append(closure)
-            domain = AUTO_DOMAIN
+
+            if extracted_lifts:
+                for tmp_sym, lift_expr in extracted_lifts.items():
+                    # make sure the applied lift is not capturing anything except of closure params
+                    used_symbol_refs = collect_symbol_refs(lift_expr)
+                    assert not (
+                        set(used_symbol_refs)
+                        - {param.id for param in current_closure_stencil.params}
+                    )
+
+                    assert isinstance(lift_expr, ir.FunCall) and isinstance(
+                        lift_expr.fun, ir.FunCall
+                    )
+
+                    # make sure the arguments to the applied lift are only symbols
+                    # (otherwise we would need to canonicalize using `canonicalize_applied_lift`
+                    # this doesn't seem to be necessary right now as we extract the lifts
+                    # in post-order of the tree)
+                    assert all(isinstance(arg, ir.SymRef) for arg in lift_expr.args)
+
+                    # create a mapping from the closures parameters to the closure arguments, e.g.
+                    # for the closure `out ← (λ(param) → ...)(arg) @ u⟨ ... ⟩;` we get a mapping
+                    # from `param` to `arg`
+                    stencil_arg_map = {
+                        param.id: arg
+                        for param, arg in zip(
+                            current_closure_stencil.params
+                            if not is_scan
+                            else current_closure_stencil.params[1:],
+                            current_closure.inputs,
+                            strict=True,
+                        )
+                    }
```

Comment:
extract function (goal is to make it possible to get an overview of this huge function/loop, by removing details)